### PR TITLE
#295 [SQ86] missing /api/alm_settings/validate

### PR DESCRIPTION
**Pull request by williamm-ic**
Added dummy /api/alm_settings/validate to SonarQube v8.6.0.39681 and up.

ALM | Pull Request Decoration | Project Import
------------ | ------------- | -------------
Azure DevOps | :heavy_check_mark: | :heavy_check_mark:
BitBucket | :heavy_check_mark: | :heavy_check_mark:
GitHub | :heavy_check_mark: | :heavy_check_mark:
GitLab | :heavy_check_mark: | :heavy_check_mark:

Download the pre-compiled jar [here](https://github.com/mc1arke/sonarqube-community-branch-plugin/files/5928334/sonarqube-community-branch-plugin-1.7.0-SNAPSHOT.bug295.SQ86.zip)

### CODE REVIEW

**Comment by benze** [comment]
@williamm-ic  I've tried the build that you posted in the top comment, but am having trouble with ALM integration.

Do you get 404 error messages for the following endpoints:
```
HTTP Status 404 : /api/alm_integrations/check_pat?almSetting=bitbucket
HTTP Status 404 : /api/alm_integrations/set_pat
```

I'm not sure if/how the ALM integration works if the /set_pat endpoint doesn't exist in the CE edition.
I created an issue #304 to track/ask if anyone else has run into this problem, but was wondering how you circumvented it.

**Comment by williamm-ic** [comment]
Hi @benze, my PR does not solve your issue. This PR is only for the missing /validate endpoint and certainly does not circumvent the issue you describe. I will investigate it when I have the time 🙂

**Comment by robmaas** [comment]
Hi @williamm-ic and @mc1arke ,
Thanks for your hard work on this, any idea when we can expect a official release for 8.6? (I'm using GitHub and Azure DevOps integration).
Cheers, Rob

**Comment by williamm-ic** [comment]
Hi @robmaas, I'm afraid I can't answer that question as I'm not a member of the team. I'm just trying to help the community.

**Comment by Atharex** [comment]
I'm having the same error as @benze when I try to import a gitlab project and I enter my PAT 
(gitlab projects cannot be imported since SQ complains about missing endpoints for gitlab alm_integration)

File: src/main/java/com/github/mc1arke/sonarqube/plugin/server/pullrequest/ws/action/ListDefinitionsAction.java [hunk 1]
```diff
@@ -69,58 +69,55 @@ public void handle(Request request, Response response) {
...
+                .build();
     }
 
     private static AlmSettings.AlmSettingGitlab toGitlab(AlmSettingDto settingDto) {
-        AlmSettings.AlmSettingGitlab.Builder almSettingBuilder =  AlmSettings.AlmSettingGitlab.newBuilder()
-            .setKey(settingDto.getKey())
-            .setPersonalAccessToken(requireNonNull(settingDto.getPersonalAccessToken(), "Personal Access Token cannot be null for Gitlab ALM setting"));
+        AlmSettings.AlmSettingGitlab.Builder almSettingBuilder = AlmSettings.AlmSettingGitlab.newBuilder()
+                .setKey(settingDto.getKey()).setPersonalAccessToken(requireNonNull(settingDto.getPersonalAccessToken(),
+                        "Personal Access Token cannot be null for Gitlab ALM setting"));
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/server/pullrequest/ws/action/ListDefinitionsAction.java:120]: The changes in this file seem to just be code formatting, which makes it difficult to tell what has actually been changed. Could you reduce your change down to just lines with functional changes please?

File: src/main/java/com/github/mc1arke/sonarqube/plugin/server/pullrequest/ws/action/ValidateAction.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2020 Michael Clarke
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/server/pullrequest/ws/action/ValidateAction.java:2]: The year should be 2021, and I'd suggest you put your name unless you're granting me ownership of your code.

**Review by mc1arke — CHANGES_REQUESTED** [review]
Thanks for the contribution. It looks like this builds on-top of one of your other PRs, but please squash any commits from either PR so that each commit delivers a feature (e.g. `SQ 8.5 support`, and `SQ 8.6 support`) rather than an initial commits and then subsequent clean-up commits.

**Comment by mc1arke** [comment]
> Hi @williamm-ic and @mc1arke ,
> Thanks for your hard work on this, any idea when we can expect a official release for 8.6? (I'm using GitHub and Azure DevOps integration).
> Cheers, Rob

Hi @robmaas 

I'll look to release this as soon as I'm happy with a code review. Unfortunately my job and personal circumstances have prevented me picking up changes to the plugin since the SQ 8.5 release hence the delays in reviewing until now, but I should be back in a position to focus on reviewing changes and working on refactoring the plugin to make some of these dummy endpoints fully functional in upcoming releases.

Thanks!

**Comment by williamm-ic** [comment]
> Thanks for the contribution. It looks like this builds on-top of one of your other PRs, but please squash any commits from either PR so that each commit delivers a feature (e.g. `SQ 8.5 support`, and `SQ 8.6 support`) rather than an initial commits and then subsequent clean-up commits.

@mc1arke
Correct, the 8.6 branch has the same codebase as the 8.5 branch and is therefore based on it. To me that seems efficient. You say that each commit should deliver a feature, but I'm not sure what you're trying to say here. Is there anything wrong with having multiple commits in a PR?

**Comment by mc1arke** [comment]
I'd expect there to be a commit for `making 8.5 work` and one for `making 8.6 work`. If you have something that causes an issue in your first attempt to do either of these commits then you should squash that commit into your original attempt.

**Comment by mc1arke** [comment]
@williamm-ic Since Sonarqube 8.7 has been released and would require removing these additional endpoints (see #310) I'm planning on closing your 2 MRs and adding notes to the README that indicates 8.5 and 8.6 are binary compatible with the last release of the plugin but are not functionally compatible for parts of Pull Request decoration.

Please let me know your thoughts.

**Comment by williamm-ic** [comment]
@mc1arke Are you still planning on releasing an update for 8.5 and 8.6? Or are you skipping them and only support 8.7?

**Comment by mc1arke** [comment]
I'm likely to skip them. Your change only adds a subset of the endpoints required to support all the ALM integrations, so the development work to make the plugin fully compatible for this historic release doesn't seem justifiable when those endpoints are immediately going to be removed for the 8.7 release.

**Comment by williamm-ic** [comment]
That is of course if everyone is willing and able to upgrade to v8.7. It's your call. Close them if you must.

**Pull request closed without merging**